lib/core/include/scipp/core/dict.h

```diff
@@ -0,0 +1,423 @@
+  }
+
+  [[nodiscard]] mapped_type &operator[](const key_type &key) {
+    return m_values[expect_find_index(key)];
+  }
+
+  [[nodiscard]] const mapped_type &at(const key_type &key) const {
+    return (*this)[key];
+  }
+
+  [[nodiscard]] mapped_type &at(const key_type &key) { return (*this)[key]; }
+
+  [[nodiscard]] const_iterator find(const key_type &key) const {
+    if (const auto key_it = find_key(key); key_it == m_keys.end()) {
+      return end();
+    } else {
+      return const_iterator(m_keys, key_it,
+                            std::next(m_values.begin(), index_of(key_it)));
+    }
+  }
+
+  [[nodiscard]] iterator find(const key_type &key) {
+    if (const auto key_it = find_key(key); key_it == m_keys.end()) {
+      return end();
+    } else {
+      return iterator(m_keys, key_it,
+                      std::next(m_values.begin(), index_of(key_it)));
+    }
+  }
+
+  [[nodiscard]] auto keys_begin() const noexcept {
+    return const_key_iterator(m_keys, m_keys.cbegin());
+  }
+
+  [[nodiscard]] auto keys_end() const noexcept {
+    return const_key_iterator(m_keys, m_keys.cend());
+  }
+
+  [[nodiscard]] auto values_begin() noexcept {
+    return value_iterator(m_values, m_values.begin());
+  }
+
+  [[nodiscard]] auto values_end() noexcept {
+    return value_iterator(m_values, m_values.end());
+  }
+
+  [[nodiscard]] auto values_begin() const noexcept {
+    return const_value_iterator(m_values, m_values.begin());
+  }
+
+  [[nodiscard]] auto values_end() const noexcept {
+    return const_value_iterator(m_values, m_values.end());
+  }
+
+  [[nodiscard]] auto begin() noexcept {
+    return iterator(m_keys, m_keys.cbegin(), m_values.begin());
+  }
+
+  [[nodiscard]] auto end() noexcept {
+    return iterator(m_keys, m_keys.cend(), m_values.end());
+  }
+
+  [[nodiscard]] auto begin() const noexcept {
+    return const_iterator(m_keys, m_keys.cbegin(), m_values.begin());
+  }
+
+  [[nodiscard]] auto end() const noexcept {
+    return const_iterator(m_keys, m_keys.cend(), m_values.begin());
+  }
+
+private:
+  Keys m_keys;
+  Values m_values;
+
+  auto find_key(const Key &key) const noexcept {
+    return std::find(m_keys.begin(), m_keys.end(), key);
+  }
+
+  auto expect_find_key(const Key &key) const {
+    if (const auto key_it = find_key(key); key_it != m_keys.end()) {
+      return key_it;
+    }
+    using scipp::core::to_string;
+    using std::to_string;
+    throw except::NotFoundError("Expected " + dict_keys_to_string(*this) +
+                                " to contain " + to_string(key) + ".");
+  }
+
+  auto index_of(const typename Keys::const_iterator &it) const noexcept {
+    return std::distance(m_keys.begin(), it);
+  }
+
+  scipp::index expect_find_index(const Key &key) const {
+    return index_of(expect_find_key(key));
+  }
+};
+
+template <class Mapping>
+std::string dict_keys_to_string(const Mapping &dict,
+                                const std::string_view dict_name = "Dict") {
```

Comment:
Can we move this to a `cpp` file? I don't like including `sstream` everywhere. Probably needs a signature change then, e.g., to `span<Key>` so we can instantiate it.

Reply:
This would require knowing all possible key types ahead of time. Do we want to lock us in like that?

Reply:
We know all possible key types. There are just two. If you are really worried, but it into a `tcc` file which potential user users can include in their `cpp`.

Reply:
Done. It is a bit tricky because `dict_keys_to_string` needs `Dict` and vice versa. So `dict_keys_to_string` is now implemented in `dict.cpp` and not `string.cpp`.